MutableIntBoundingBox -> BlockBox

Sollace opened the pull request:
Last PR for today!

I renamed `MutableIntBoundingBox` to `MutableBox` as per #785. I also standardised the naming between the two clases.

`x0, y0, z0, x1, y1, z1` -> `minX, minY, minZ, maxX, maxY, maxZ`

I also kept the names `getXSize`, `getYSize`, and `getZSize` for now, though I want to know what people would think of me changing them to `getWidth`, `getHeight`, and `getDepth` instead.

Sollace committed:
MutableIntBoundingBox -> MutableBox

Pyrofab commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -32,6 +38,14 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
 	METHOD method_1005 getCenter ()Lnet/minecraft/class_243;
 	METHOD method_1006 contains (Lnet/minecraft/class_243;)Z
 		ARG 1 vec
+	METHOD method_1007 traceCollissionSide (Lnet/minecraft/class_238;Lnet/minecraft/class_243;[DLnet/minecraft/class_2350;DDD)Lnet/minecraft/class_2350;
```
`traceCollisionSide`
Also, what does this method do exactly ? I can't really tell just by looking at the name.

Sollace replied:
Best I can say is it it's used for rayTracing to find either the side of the box the ray enters, or that the ray exits.

It's called once for each side, every time passing in a shared double[] array and the direction produced by the last call of the method.

    @Nullable
    public static BlockHitResult rayTrace(Iterable<Box> boxes, Vec3d from, Vec3d to, BlockPos blockPos) {
        double[] doubles_1 = { 1.0 };
        Direction direction6 = null;
        double double7 = to.x - from.x;
        double double9 = to.y - from.y;
        double double11 = to.z - from.z;
        for (final Box class_238_1 : boxes) {
            direction6 = traceCollissionSide(class_238_1.offset(blockPos), from, doubles_1, direction6, double7, double9, double11);
        }
        if (direction6 == null) {
            return null;
        }
        double double13 = doubles_1[0];
        return new BlockHitResult(from.add(double13 * double7, double13 * double9, double13 * double11), direction6, blockPos, false);
    }
    
    @Nullable
    private static Direction traceCollissionSide(Box box, Vec3d approachVector, double[] traceDistanceResult, @Nullable Direction approachDirection, double xDelta, double yDelta, double zDelta) {
        if (xDelta > 1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, xDelta, yDelta, zDelta, box.minX, box.minY, box.maxY, box.minZ, box.maxZ, Direction.WEST, approachVector.x, approachVector.y, approachVector.z);
        }
        else if (xDelta < -1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, xDelta, yDelta, zDelta, box.maxX, box.minY, box.maxY, box.minZ, box.maxZ, Direction.EAST, approachVector.x, approachVector.y, approachVector.z);
        }
        if (yDelta > 1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, yDelta, zDelta, xDelta, box.minY, box.minZ, box.maxZ, box.minX, box.maxX, Direction.DOWN, approachVector.y, approachVector.z, approachVector.x);
        }
        else if (yDelta < -1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, yDelta, zDelta, xDelta, box.maxY, box.minZ, box.maxZ, box.minX, box.maxX, Direction.UP, approachVector.y, approachVector.z, approachVector.x);
        }
        if (zDelta > 1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, zDelta, xDelta, yDelta, box.minZ, box.minX, box.maxX, box.minY, box.maxY, Direction.NORTH, approachVector.z, approachVector.x, approachVector.y);
        }
        else if (zDelta < -1.0E-7) {
            approachDirection = traceCollissionSide(traceDistanceResult, approachDirection, zDelta, xDelta, yDelta, box.maxZ, box.minX, box.maxX, box.minY, box.maxY, Direction.SOUTH, approachVector.z, approachVector.x, approachVector.y);
        }
        return approachDirection;
    }
    
    @Nullable
    private static Direction traceCollissionSide(double[] traceDistanceResult, @Nullable Direction approachDirection, double xDelta, double yDelta, double zDelta, double begin, double minX, double maxX, double minZ, double maxZ, Direction resultDirection, double startX, double startY, double startZ) {
        double double26 = (begin - startX) / xDelta;
        double double28 = startY + double26 * yDelta;
        double double30 = startZ + double26 * zDelta;
        if (0.0 < double26 && double26 < traceDistanceResult[0] && minX - 1.0E-7 < double28 && double28 < maxX + 1.0E-7 && minZ - 1.0E-7 < double30 && double30 < maxZ + 1.0E-7) {
            traceDistanceResult[0] = double26;
            return resultDirection;
        }
        return approachDirection;
    }

Pyrofab commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -32,6 +38,14 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
 		ARG 1 vec
+	METHOD method_1007 traceCollissionSide (Lnet/minecraft/class_238;Lnet/minecraft/class_243;[DLnet/minecraft/class_2350;DDD)Lnet/minecraft/class_2350;
+		ARG 0 box
+		ARG 1 approachVector
```
There may be a more appropriate term than "approach". Does `intersectingVector` make sense in this context ?

Sollace replied:
`intersectionVector` works better, yes.

Pyrofab commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -58,23 +72,44 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
 		ARG 1 vec3d
-	METHOD method_998 ([DLnet/minecraft/class_2350;DDDDDDDDLnet/minecraft/class_2350;DDD)Lnet/minecraft/class_2350;
-		ARG 18 maxZ
+	METHOD method_998 traceCollissionSide ([DLnet/minecraft/class_2350;DDDDDDDDLnet/minecraft/class_2350;DDD)Lnet/minecraft/class_2350;
```
traceCollisionSide

Pyrofab commented on mappings/net/minecraft/util/math/MutableBox.mapping:
```diff
@@ -28,23 +28,23 @@ CLASS net/minecraft/class_3341 net/minecraft/util/math/MutableIntBoundingBox
+		ARG 3 maxX
+		ARG 4 maxY
+		ARG 5 maxZ
 	METHOD method_14667 createRotated (IIIIIIIIILnet/minecraft/class_2350;)Lnet/minecraft/class_3341;
```
We could consider renaming this to `rotated` maybe ?

Sollace replied:
This returns a new instance with the rotation, rather than this one. I don't feel simply calling it `rotated` would adequately portray that.

Pyrofab replied:
`rotate` would be the mutating method IMO

Sollace replied:
Again, this method does not mutate the box. Returns a new box.

JamiesWhiteShirt replied:
Which `rotated` adequately portrays in my opinion.

Pyrofab replied:
`rotate` != `rotated`
Rotate implies mutation to a rotated state, rotated implies the creation of a rotated copy

Pyrofab commented:
I think `getWidth`, `getHeight` and `getDepth` sound much better than `getXYZSize`.

Sollace committed:
collission -> collision

Sollace committed:
approachVector -> intersectingVector

Sollace commented:
@Pyrofab If I can get more people to agree, I'll put it in.

Another related thing I meant to ask is:
What becomes width, and what becomes depth?

I know `y = height`, but `must x = width` and `z =depth` or must it be `x = depth` and `z = width`?

Pyrofab commented:
Seeing how width and depth are not immediately clear, I have reconsidered and I would suggest `getXYZLength` instead.

Sollace committed:
createRotated -> rotated (and map parameters)

Pyrofab commented on mappings/net/minecraft/util/math/MutableBox.mapping:
```diff
@@ -28,27 +28,33 @@ CLASS net/minecraft/class_3341 net/minecraft/util/math/MutableIntBoundingBox
 		ARG 1 other
 	METHOD method_14658 toNbt ()Lnet/minecraft/class_2495;
 	METHOD method_14659 getSize ()Lnet/minecraft/class_2382;
```
New suggestion: `getSize` -> `getDimensions`. The latter is technically more correct and disambiguates it from the `getXYZSize` methods.

Sollace replied:
So `getXDimention`?

I haven't looked at this for a few days, but iirc `getSize` here returns a scalar value, so perhaps `getVolume` would be better?

Pyrofab replied:
Nop, keep `getXSize`, but rename the base `getSize` to `getDimensions`. That is because sizes are single values, whereas dimensions are vectors.

Runemoro commented:
I don't see why width, height, and depth would be x, y, z and not z, y, x... 

I suggest using `getSizeX` instead of `getXSize` to respect [the convention](https://github.com/FabricMC/yarn/blob/1.14.4/CONVENTIONS.md#coordinates) (it's the size's x component, not the "x size").

Sollace committed:
getSize -> getDimensions

Sollace committed:
size -> length

Sollace commented:
Whilst looking at MurableBox, I notice that the constructor not specifying a y range results in a box defaulted to 0-256 (world height).

This class is also really only used with structures and commands, so it makes me wonder...

This could either just be a `SelectionBox` or a `BlockSelectionArea`, or some mix of the two - something that represents an area in the world rather than entity hitboxes* like the oher one is used for.

* inb4 `Box` -> `Hitbox`

Pyrofab commented:
`Box` is used for a lot more than hitboxes, and MutableBox uses the [1, 512] y range by default. However, the fact that the latter uses whole numbers and that it is only used for blocks could be taken into account for naming it. I am thinking `BlockBox` may be a fairly straightforward name, in the same vein as `BlockPos`.

Sollace committed:
MutableBox -> BlockBox

Runemoro commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -54,27 +68,48 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
-	METHOD method_17941 getZSize ()D
+	METHOD method_17939 getXLength ()D
+	METHOD method_17940 getYLength ()D
+	METHOD method_17941 getZLength ()D
```
Why change this? Every other size getter uses `getSizeX`/`Y`/`Z`. Using `length` in this one would be weird.

Pyrofab replied:
This change was originally a suggestion of mine, but I agree that we should probably stick to established conventions.

Sollace replied:
`getSizeX` IMO, is counter to the CONVENTIONS.md

> To make code as easy to read as possible, keep names in the natural language order. For example, a class representing a chest block entity should be named `ChestBlockEntity` rather than `BlockEntityChest`.

Runemoro replied:
What? `getXSize` breaks the convention. There's no such thing as an "x size" (unless you're talking about the size of the "x" character in a font). There is though a "size's x", the x component of the size.

Also, the conventions mention this explicitly: https://github.com/FabricMC/yarn/blob/1.14.4/CONVENTIONS.md#coordinates

In any case, even if you disagree, it should still be `sizeX`/`Y`/`Z` and you should open an issue about changing *all* `somethingX`/`Y`/`Z`/ variables to `x`/`y`/`zSomething`, not just change this one and be inconsistent ([within the same class!](https://github.com/FabricMC/yarn/pull/801/files#diff-281a29125b212111a67a0e63eb6dfaffR31)). That would make people have to remember which methods have the coordinate in front and which have it at the end.

Sollace replied:
Get X Size = get the size in the X dimension
Get X Length = get the length of the X component

Compare that to a real conversaion:

"What is your cup('s) size?"
"My cup('s) size is DD"

As opposed to:

"What is your size('s) cup?"
"My size has a cup?"

Do you even english?

Sollace replied:
You can also consider them a mutation of `x`, `y`, and `z`.

Where each component is a scalar _range_. If the mins and maxes of each dimension where refactored into a dedicated `Range` object the equivalent of these methods would all be:
`getX().length()`, `getY().length()`, `getZ().length()`.

Or.

`getX().getSize()`, `getY().getSize()`, `getZ().getSize()`.

Runemoro replied:
I have never heard anyone refer to the size in the x dimension as the "x size". 

> What is your size('s) cup?

A size doesn't have a cup. But a cup has a size, and that size has three components.

A size doesn't have a box. But a box has a size, and that size has three components.

> Do you even english?

Do you even logic?

Also, ironically, you forgot the capital letter in English exactly in the sentence you insult my English.

> Where each component is a scalar range. If the mins and maxes of each dimension where refactored into a dedicated Range object the equivalent of these methods would all be:
getX().length(), getY().length(), getZ().length().

I would never guess that a method named `getX` returns a range, I'd instead assume it returns the position on the x axis. If I saw `getX().length()` I'd be confused about how a position can possibly have a length, just as I'd be confused when I see `getXLength`.

If you really want to use the (weird) range naming, then it would have to be `getXRangeLength` (get the x [coordinate's] range length) to be clear.

Also, by the same argument as yours, I can argue that `getSizeX` is correct, because it represents `getSize().getX()`. It's obviously more common to see methods that return sizes than methods that return ranges, so this would be the better option even if the other one was correct (not saying it is).

I suggest that this PR doesn't get merged as it is until a PR changing `CONVENTIONS.MD` is merged, since it's currently breaking the conventions.

Sollace replied:
I'll agree to that until someone else can weigh in on this.

As it is, I'm not against renaming the methods to something else. Just as it stands, though the pre-X version are _obviously_ better in my opnion, neither options `getSizeX`, `getLengthX`, `getXSize` or `getXLength` are very elegant names due to the mixing of "dimension-axis + arbitrary-modifier".

Runemoro commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -54,27 +68,48 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
 	METHOD method_994 intersects (Lnet/minecraft/class_238;)Z
 		ARG 1 box
-	METHOD method_995 averageDimension ()D
+	METHOD method_995 getAverageSideLength ()D
```
Why change this? "Dimension" means "side length" but is more concise.

> [a measurable extent of some kind, such as **length, breadth, depth, or height**](https://www.google.com/search?q=dimension)

Pyrofab replied:
I think "side length" is clearer, but that's really a matter of taste.

Sollace replied:
I'm with Pyro on this one. Not only that, but `dimension` is very similar to `world dimension`, `entity dimensions`, and just the term itself implies this method returns a vector, or therwise multiple scalars to describe the "dimensions" of the box. (see: `BlockBox.getDimensions`)

`Length` on the other hand, is always a scalar.

Runemoro commented on mappings/net/minecraft/util/math/BlockBox.mapping:
```diff
@@ -1,4 +1,4 @@
-CLASS net/minecraft/class_3341 net/minecraft/util/math/MutableIntBoundingBox
+CLASS net/minecraft/class_3341 net/minecraft/util/math/BlockBox
```
We already have a class named `Box` which it's just an immutable version of this one (the same way we have `BlockPos` and `MutableBlockPos`). Either this class should be `MutableBox` and `Box` should stay `Box`, or this class should be `MutableBlockBox` and `Box` should be named `BlockBox`.

Pyrofab replied:
The idea behind calling this one `BlockBox` is that it's the equivalent of the `Box` class but for blocks - it uses ints and is used only in block contexts. The main difference between the other `Box` and this one is not mutability, it's the type of data it contains.

Runemoro replied:
> The main difference between the other Box and this one is not mutability, it's the type of data it contains.

They contain exactly the same data (ints minX, minY, minZ, maxX, maxY, maxZ) and have the same methods.

Pyrofab replied:
Not quite, the original `Box` uses doubles, and is used in a lot of contexts. This one uses ints and is limited to block contexts.

Runemoro replied:
> the original Box uses doubles,

Sorry, I thought it was also using ints, I should have looked at the code.

Sollace replied:
So... "conversation resolved"?

Runemoro commented on mappings/net/minecraft/util/math/BlockBox.mapping:
```diff
@@ -27,28 +27,34 @@ CLASS net/minecraft/class_3341 net/minecraft/util/math/MutableIntBoundingBox
-	METHOD method_14663 getBlockCountY ()I
-	METHOD method_14664 getBlockCountZ ()I
+	METHOD method_14663 getYLength ()I
+	METHOD method_14664 getZLength ()I
```
I think it should stay block count to avoid confusion between "length" and "dimensions". The `getDimensions` method returns max - min:

    public Vec3i getSize() {
        return new Vec3i(this.maxX - this.minX, this.maxY - this.minY, this.maxZ - this.minZ);
    }

while the `getBlockCount` methods return max - min + 1 (the number of blocks, so a box from (0, 0, 0) to (0, 0, 0) will have a block count of 1 since it contains the block at (0, 0, 0):

    public int getBlockCountX() {
        return this.maxX - this.minX + 1;
    }
    
    public int getBlockCountY() {
        return this.maxY - this.minY + 1;
    }
    
    public int getBlockCountZ() {
        return this.maxZ - this.minZ + 1;
    }

Sollace replied:
Yeah, I think so. It was originally changed to `getX/Y/ZSize` to conform with the method in `Box` but now that the class has been renamed to `BlockBox` there is no longer any need to maintain parity.

Sollace replied:
The names are _horribly_ unwieldy and clunky to read, though, so I welcome any, literally _any_ other suggestions.

Pyrofab replied:
I reckon this can always be changed later if someone has a better idea.

Sollace committed:
getX/Y/ZLength <- getBlockCountX/Y/Z

valoeghese commented:
*cue all structure mods causing syntax errors in the dev environment*

Runemoro commented:
Is anything stopping this PR from being merged? It's been open for almost a month and has two approvals since more than a week.

NinjaPhenix commented on mappings/net/minecraft/util/math/MutableBox.mapping:
```diff
@@ -28,23 +28,23 @@ CLASS net/minecraft/class_3341 net/minecraft/util/math/MutableIntBoundingBox
+		ARG 2 minZ
+		ARG 3 maxX
+		ARG 4 maxY
+		ARG 5 maxZ
```
Why was this change to min/max when coords can be given in either position? e.g. `create(1, 1, 1, 2, 2, 2)` and `create(2, 2, 2, 1, 1, 1)` results in same box.
```java
public static MutableIntBoundingBox create(int int_1, int int_2, int int_3, int int_4, int int_5, int int_6) {
    return new MutableIntBoundingBox(Math.min(int_1, int_4), Math.min(int_2, int_5), Math.min(int_3, int_6), Math.max(int_1, int_4), Math.max(int_2, int_5), Math.max(int_3, int_6));
}
```

Sollace replied:
Because when I was working on this the general consensus was against using x/y/z1 and x/y/z2. These names are also not very friendly, and actually rather confusing as I've found in the past.

Runemoro replied:
@Sollace I agree with @NinjaPhenix, these should be changed back to `x1`, `x2`. The new names are much more confusing, they're making me think I have to make sure they are min and max, when they can in fact be max and min. Please don't add extra restrictions through parameter names.

> and actually rather confusing as I've found in the past.

What's confusing about being able to make a box between two points, with no restrictions at all...?


@liach @Pyrofab (mentioning you in case like me you weren't aware of this when you reviewed) What are your thoughts?

Pyrofab replied:
Agreed, at first I thought the `create` method just passed everything to the constructor without changing the argument order.

NinjaPhenix commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -6,11 +6,17 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
+		ARG 5 minZ
+		ARG 7 maxX
+		ARG 9 maxY
+		ARG 11 maxZ
```
Same as my other comment, positions can be given in either order so why use min/max.

NinjaPhenix replied:
Other occurances exist in BlockPos, vec3d constructors aswell.

Pyrofab replied:
According to the code you posted above, the `create` methods put the values in the right order for the constructor. Meaning the constructor does care about the order, and the argument names are correct.

Runemoro reviewed (changes requested):
(see https://github.com/FabricMC/yarn/pull/801#discussion_r317400638)

liach commented on mappings/net/minecraft/util/math/Box.mapping:
```diff
@@ -6,11 +6,17 @@ CLASS net/minecraft/class_238 net/minecraft/util/math/Box
+		ARG 5 minZ
+		ARG 7 maxX
+		ARG 9 maxY
+		ARG 11 maxZ
```
Given mojang does not require max/min, consider names like `xa` `xb` etc.

Sollace committed:
min/max -> pos1/pos2

Sollace committed:
min/max -> pos1/pos2 (constructor still requires min/max)

Sollace commented:
Alright, changed back to pos1/pos2 then (bleh)

Only outlier is the constructor for BlockBox

```
    public BlockBox(int minX, int minY, int minZ, int maxX, int maxY, int maxZ) {
        this.minX = minX;
        this.minY = minY;
        this.minZ = minZ;
        this.maxX = maxX;
        this.maxY = maxY;
        this.maxZ = maxZ;
    }
```
(this is why I changed them, so it's standardized between all the methods)

Sollace committed:
setFrom -> encompass (this expands the current selection to encompass the passed region)

Sollace commented:
@Runemoro Might have to re-re-approve this. I don't know.
